cpp/harvest/core.cc

```diff
@@ -318,17 +318,39 @@ void Filter(int argc, char **argv) {
     // std::sort(filter_data_provider_ids.begin(), filter_data_provider_ids.end());
 
     for (auto work : works) {
-        if (not filter_data_provider_ids.empty()) {
-            auto sortedDataProvider = work.getDataProviderIds();
-            const bool is_in(MiscUtil::Intersect(sortedDataProvider, filter_data_provider_ids).size() > 0);
-            if (is_in == filter) {
-                work.setFilteredReason("Data Provider");
-                CORE::OutputFileAppend(output_file_skip, work, skipped == 0);
-                ++skipped;
-                ++skipped_data_provider_count;
-                continue;
+        if (argc == 7) {
+            if (!filter) {
+                // this means keep the record but clean the member of data provider
+                // clean the member of data provider means delete the member that its id is not in the list
+                // this is 'keep' option
+                auto data_provider_ids = work.getDataProviderIds();
+                const bool is_in_the_list(MiscUtil::Intersect(data_provider_ids, filter_data_provider_ids).size() > 0);
+
+                if (!data_provider_ids.empty() && is_in_the_list) {
+                    work.removeDataProviders(filter_data_provider_ids);
```

Comment:
You can use the the intersection result as input for `removeDataProviders()` instead for better performance. Rest looks good!